blockchain/My-Wallet-V3-iOS
# feat(KYC): New Address Screen & ValidationTextField

## Objective

This updates the address entry screen to reflect the most recent designs as well as introduces `ValidationTextField`. Previous classes such as `CellModel` and the `KYCAddressDetailsViewController` have been removed.

## Description

The address screen is now composed of:
* `UISearchBar`
* `UIProgressView`
* `UITableView`
* `UIScrollView` - Contains multiple `ValidationTextFields` and the `primaryButton`

`ValidationTextField` is what we will use to indicate to the user whether or not the values they provide to a specific `UITextField` pass certain requirements. `ValidationTextField` is a subclass of `NibBasedView`. This makes adding `ValidationTextField` to a storyboard a bit easier.

Many of the properties on the class are `@IBInspectable` so when adding this to your screen, be sure to use IB where you can. Some properties you can't use IB for such as setting the `UIKeyboardType`, the `UIReturnKeyType`, or the font. There are closures provide for handling custom logic:
* `returnTappedBlock` - Used for routing the user to the next `ValidationTextField`
* `becomeFirstResponderBlock` - Used for handling scrolling events when the field is activated.
* `validationBlock` - Used for implementing custom validation logic (e.g. phone number requirements, email validation, etc.)
You can set `optionalField` to `false` to make the field required. 

`KYCAddressController` now includes some logic for handling the offsetting of the `UIScrollView` to make sure the currently active input field is within the field of view. 

## How to Test

1. Build and run
2. Tap `Debug` and go into the KYC flow. 
**Note:** You may get blocked by the phone number verification screen. If so, comment out the line where the phoneNumber field is the first responder, and add `handleSendVerificationCodeSuccess()` and a `return` in the `verify(number: String, userId: String)` function in `KYCVerifyPhoneNumberPresenter`.
3. Once you get to the address screen, search for an address
4. Select an address
5. The fields should autopopulate and you should be able to continue through the flow. 
6. Delete one of the fields that is *required* and then try to tap continue
7. The field should show an error state and the field should become the first responder. 

## Screenshot

![screen shot 2018-08-03 at 10 38 46 am](https://user-images.githubusercontent.com/41585563/43653922-2c98e920-970f-11e8-9508-9338e9d7c10e.png)
![screen shot 2018-08-03 at 10 43 09 am](https://user-images.githubusercontent.com/41585563/43653930-324b5178-970f-11e8-8626-86cd0bcaf4f6.png)

## Related Information

* Ticket Number: IOS-1134
* Ticket Number: IOS-1115

## Merge Checklist

- [x] New files added are within the correct directory.
